### PR TITLE
fix: Prevent adding ownership condition for NotFound objects

### PR DESCRIPTION
**Pull request by cowsking**
This commit fixes a bug in the ResourceGroup controller where the ownership condition was occasionally added to the status for objects which were not found.

This happened when the object was cached as NotFound by the FilteredWatcher. The Reconcile loop would then use the cached object and call setResStatus, which incorrectly added the ownershipCondition.

The fix ensures that if the cached status is NotFound, the reconciler skips the call to setResStatus.

A unit test, TestComputeStatus, has been added to verify this behavior.

### CODE REVIEW

File: pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller.go [hunk 1]
```diff
@@ -283,6 +283,10 @@ func (r *reconciler) computeStatus(
 		switch {
 		case cachedStatus != nil:
 			log.V(4).Info("Resource object status found in the cache")
+			if cachedStatus.Status == v1alpha1.NotFound {
+				resStatus.Status = v1alpha1.NotFound
+				break
+			}
 			setResStatus(id, &resStatus, cachedStatus)
```
[the remark this answers is not in the document]
Comment (mikebz) [reply pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller.go]: Agreed with this feedback.

[the remark this answers is not in the document]
Comment (sdowell) [reply pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller.go]: +1 also agree with this feedback

**Comment by cowsking** [comment]
> @cowsking: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> kpt-config-sync-presubmit-e2e-multi-repo-test-group1	[c0678a9](https://github.com/GoogleContainerTools/kpt-config-sync/commit/c0678a9122d587a6c3e9939df6afc01f7dc105b1)	[link](https://oss.gprow.dev/view/gs/oss-prow-build-kpt-config-sync/pr-logs/pull/GoogleContainerTools_kpt-config-sync/1819/kpt-config-sync-presubmit-e2e-multi-repo-test-group1/1955006965339918336)	true	`/test kpt-config-sync-presubmit-e2e-multi-repo-test-group1`

/retest

File: pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (cowsking) [reply pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller.go]: @sdowell @mikebz The suggestion has been committed

File: pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller_test.go [hunk 2]
```diff
@@ -803,3 +805,43 @@ func TestUpdateStatusToReflectActuation(t *testing.T) {
...
+		Status: v1alpha1.NotFound,
+	}
+	resMap.SetStatus(objMeta, cachedStatus)
+
+	// 3. Call computeStatus.
+	// This is the "Act" step of the test.
+	resultStatuses := r.computeStatus(context.Background(), inventoryID, v1alpha1.ResourceGroupStatus{}, []v1alpha1.ObjMetadata{objMeta}, types.NamespacedName{}, true)
+
+	// 4. Assert the results.
+	require.Len(t, resultStatuses, 1, "Expected exactly one resource status to be returned")
```
Comment (sdowell) [inline pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller_test.go]: This can be simplified to a single assertion:
```golang
	wantStatus := []v1alpha1.ResourceStatus{
		{
			ObjMetadata: objMeta,
			Status:      v1alpha1.NotFound,
		},
	}
	require.Equal(t, wantStatus, resultStatuses)
```

File: pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller_test.go [hunk 3]
```diff
@@ -803,3 +805,43 @@ func TestUpdateStatusToReflectActuation(t *testing.T) {
...
+	// This simulates the scenario where the FilteredWatcher has already determined
+	// the object does not exist.
+	cachedStatus := &resourcemap.CachedStatus{
+		Status: v1alpha1.NotFound,
+	}
+	resMap.SetStatus(objMeta, cachedStatus)
+
+	// 3. Call computeStatus.
+	// This is the "Act" step of the test.
+	resultStatuses := r.computeStatus(context.Background(), inventoryID, v1alpha1.ResourceGroupStatus{}, []v1alpha1.ObjMetadata{objMeta}, types.NamespacedName{}, true)
```
Comment (sdowell) [inline pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller_test.go]: use `t.Context()` instead of `context.Background()`

File: pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller_test.go [hunk 4]
```diff
@@ -803,3 +805,43 @@ func TestUpdateStatusToReflectActuation(t *testing.T) {
 		})
 	}
 }
+
+func TestComputeStatus_CachedNotFoundSkipsOwnership(t *testing.T) {
```
Comment (sdowell) [inline pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller_test.go]: I'd suggest structuring this as a more generic data driven test, as that will make this unit test more scalable/maintainable. 

While this test as-is successfully validates this one edge case, we could better improve the overall test health here by delivering some more test coverage at the same time. We could add some additional test cases to improve the test coverage of this function, such as:
- computeStatus for an object that is cached as not found (current test)
- computeStatus for an object that is cached and exists
- computeStatus for an object that is not cached and is not found
- computeStatus for an object that is not cached and exists
- etc

It's generally a good practice to leave the code a little bit better than how you found it 😄

Comment (sdowell) [reply pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller_test.go]: Suggest writing this as a data/table driven test. See:
- https://go.dev/wiki/TableDrivenTests
- go/go-style/decisions#table-driven-tests
- go/gotip/episodes/21

File: pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller_test.go [hunk 5]
```diff
@@ -30,13 +30,16 @@ import (
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/watch"
+	discoveryfake "k8s.io/client-go/discovery/fake"
 	"k8s.io/client-go/rest"
+	fakediscovery "k8s.io/client-go/testing"
```
Comment (sdowell) [inline pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller_test.go]: suggest `clientgotesting`

File: pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller_test.go [hunk 6]
```diff
@@ -30,13 +30,16 @@ import (
...
 	"k8s.io/klog/v2"
 	"kpt.dev/configsync/pkg/api/kpt.dev/v1alpha1"
 	"kpt.dev/configsync/pkg/metadata"
+	"kpt.dev/configsync/pkg/reconcilermanager/controllers"
 	"kpt.dev/configsync/pkg/resourcegroup/controllers/resourcemap"
 	"kpt.dev/configsync/pkg/resourcegroup/controllers/typeresolver"
-	"kpt.dev/configsync/pkg/syncer/syncertest/fake"
+	syncertestfake "kpt.dev/configsync/pkg/syncer/syncertest/fake"
```
Comment (sdowell) [inline pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller_test.go]: This doesn't need an alias

File: pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller_test.go [hunk 7]
```diff
@@ -803,3 +806,165 @@ func TestUpdateStatusToReflectActuation(t *testing.T) {
...
+			resMap := resourcemap.NewResourceMap()
+			for objMeta, cachedStatus := range tc.cachedStatuses {
+				resMap.SetStatus(objMeta, cachedStatus)
+			}
+
+			discoveryClient := &discoveryfake.FakeDiscovery{Fake: &fakediscovery.Fake{}}
+			discoveryClient.Resources = tc.discoveryResources
+			resolver := typeresolver.NewTypeResolver(discoveryClient, testLogger)
+			if tc.discoveryResources != nil {
+				require.NoError(t, resolver.Refresh(context.Background()))
```
Comment (sdowell) [inline pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller_test.go]: use `t.Context()` instead of `context.Background()` here and below

File: pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller_test.go [hunk 8]
```diff
@@ -803,3 +806,165 @@ func TestUpdateStatusToReflectActuation(t *testing.T) {
...
+			r := &reconciler{
+				LoggingController: controllers.NewLoggingController(testLogger),
+				client:            fakeClient,
+				resolver:          resolver,
+				resMap:            resMap,
+			}
+
+			resultStatuses := r.computeStatus(context.Background(), inventoryID, v1alpha1.ResourceGroupStatus{}, tc.objMetas, types.NamespacedName{}, tc.useCache)
+
+			opts := []cmp.Option{
```
Comment (sdowell) [inline pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller_test.go:957]: Is this needed? It doesn't look like any of the test inputs have conditions

File: pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller_test.go [hunk 9]
```diff
@@ -803,3 +806,165 @@ func TestUpdateStatusToReflectActuation(t *testing.T) {
...
+		Namespace: "default",
+		GroupKind: v1alpha1.GroupKind{Kind: "Pod", Group: ""},
+	}
+
+	testCases := []struct {
+		name               string
+		initialObjects     []client.Object
+		objMetas           []v1alpha1.ObjMetadata
+		cachedStatuses     map[v1alpha1.ObjMetadata]*resourcemap.CachedStatus
+		useCache           bool
```
Comment (sdowell) [inline pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller_test.go]: this variable doesn't seem like it's needed. whether to use the cache is implied by the `cachedStatuses` variable
